drv/sidecar-seq-server/src/main.rs

```diff
@@ -24,177 +29,335 @@ use i2c_config::devices;
+    TofinoSequencerTick(TofinoSequencerPolicy, TofinoSeqState, TofinoSeqError),
+    TofinoSequencerError(SeqError),
+    TofinoSequencerFault(TofinoSeqError),
+    TofinoVidAck,
+    InitiateTofinoPowerUp,
+    InitiateTofinoPowerDown,
+    SetVddCoreVout(userlib::units::Volts),
+    ClearingTofinoSequencerFault(TofinoSeqError),
 }
 
-ringbuf!(Trace, 64, Trace::None);
+ringbuf!(Trace, 32, Trace::None);
 
-const TIMER_MASK: u32 = 1 << 0;
+const TIMER_NOTIFICATION_MASK: u32 = 1 << 0;
 const TIMER_INTERVAL: u64 = 1000;
 
-struct ServerImpl {
-    state: PowerState,
-    clockgen: I2cDevice,
-    led: drv_stm32xx_sys_api::PinSet,
-    led_on: bool,
-    deadline: u64,
-    clock_config_loaded: bool,
+struct ClockGenerator {
+    device: I2cDevice,
+    config_loaded: bool,
 }
 
-impl ServerImpl {
-    fn led_init(&mut self) {
-        use drv_stm32xx_sys_api::*;
+impl ClockGenerator {
+    fn load_config(&mut self) -> Result<(), SeqError> {
+        ringbuf_entry!(Trace::LoadingClockConfiguration);
 
-        let sys = SYS.get_task_id();
-        let sys = Sys::from(sys);
+        let mut packet = 0;
 
-        // Make the LED an output.
-        sys.gpio_configure_output(
-            self.led,
-            OutputType::PushPull,
-            Speed::High,
-            Pull::None,
-        )
-        .unwrap();
+        payload::idt8a3xxxx_payload(|buf| match self.device.write(buf) {
+            Err(err) => {
+                ringbuf_entry!(Trace::ClockConfigurationError(packet, err));
+                Err(SeqError::ClockConfigurationFailed)
+            }
+
+            Ok(_) => {
+                packet += 1;
+                Ok(())
+            }
+        })?;
+
+        self.config_loaded = true;
+        Ok(())
     }
+}
 
-    fn led_on(&mut self) {
-        use drv_stm32xx_sys_api::*;
+struct Tofino {
+    policy: TofinoSequencerPolicy,
+    sequencer: TofinoSequencer,
+    vddcore: Raa229618,
+}
 
-        let sys = SYS.get_task_id();
-        let sys = Sys::from(sys);
-        sys.gpio_set_to(self.led, true).unwrap();
-        self.led_on = true;
+impl Tofino {
+    fn apply_vid(&mut self, vid: Tofino2Vid) -> Result<(), SeqError> {
+        use userlib::units::Volts;
+
+        let mut set_vout = |value: Volts| {
+            self.vddcore
+                .set_vout(value)
+                .map_err(|_| SeqError::SetVddCoreVoutFailed)?;
+
+            ringbuf_entry!(Trace::SetVddCoreVout(value));
+            Ok(())
+        };
+
+        match vid {
+            Tofino2Vid::V0P922 => set_vout(Volts(0.922)),
+            Tofino2Vid::V0P893 => set_vout(Volts(0.893)),
+            Tofino2Vid::V0P867 => set_vout(Volts(0.867)),
+            Tofino2Vid::V0P847 => set_vout(Volts(0.847)),
+            Tofino2Vid::V0P831 => set_vout(Volts(0.831)),
+            Tofino2Vid::V0P815 => set_vout(Volts(0.815)),
+            Tofino2Vid::V0P790 => set_vout(Volts(0.790)),
+            Tofino2Vid::V0P759 => set_vout(Volts(0.759)),
+        }
```

Comment:
Nit: this is a little over-complicated, how about
```suggestion
        let value = Volts(match vid {
            Tofino2Vid::V0P922 => 0.922,
            Tofino2Vid::V0P893 => 0.893,
            Tofino2Vid::V0P867 => 0.867,
            Tofino2Vid::V0P847 => 0.847,
            Tofino2Vid::V0P831 => 0.831,
            Tofino2Vid::V0P815 => 0.815,
            Tofino2Vid::V0P790 => 0.790,
            Tofino2Vid::V0P759 => 0.759,
        });
        self.vddcore
            .set_vout(value)
            .map_err(|_| SeqError::SetVddCoreVoutFailed)?;

        ringbuf_entry!(Trace::SetVddCoreVout(value));
        Ok(())
```

Reply:
This is significantly cleaner. Thanks!